storage/include/rocksdb/native_client.h

```diff
@@ -0,0 +1,542 @@
+  void dropColumnFamily(const std::string &cFamily);
+
+ private:
+  NativeClient(const std::string &path, bool readOnly);
+
+  // Make sure we only allow types that have data() and size() members. That excludes raw pointers without corresponding
+  // size.
+  template <
+      typename Span,
+      std::enable_if_t<std::is_convertible_v<decltype(std::declval<Span>().size()), std::size_t> &&
+                           std::is_pointer_v<decltype(std::declval<Span>().data())> &&
+                           std::is_convertible_v<std::remove_pointer_t<decltype(std::declval<Span>().data())> (*)[],
+                                                 const char (*)[]>,
+                       int> = 0>
+  static ::rocksdb::Slice toSlice(const Span &span) noexcept {
+    return ::rocksdb::Slice{span.data(), span.size()};
+  }
+
+  static void throwOnError(std::string_view msg1, std::string_view msg2, ::rocksdb::Status &&);
+  static void throwOnError(std::string_view msg, ::rocksdb::Status &&);
+
+  // For use in column family unique pointers that are managed solely by NativeClient. This allows us to use a raw
+  // pointer in CfDeleter as we know the column family unique pointer will not be moved out of NativeClient.
+  struct CfDeleter {
+    void operator()(::rocksdb::ColumnFamilyHandle *h) const noexcept {
+      if (!client_->dbInstance_->DestroyColumnFamilyHandle(h).ok()) {
+        std::terminate();
+      }
+    }
+    Client *client_{nullptr};
+  };
+
+  using CfUniquePtr = std::unique_ptr<::rocksdb::ColumnFamilyHandle, CfDeleter>;
+
+  ::rocksdb::ColumnFamilyHandle *defaultColumnFamilyHandle() const;
+  ::rocksdb::ColumnFamilyHandle *columnFamilyHandle(const std::string &cFamily) const;
+  CfUniquePtr createColumnFamilyHandle(const std::string &cFamily);
+
+ private:
+  std::string path_;
+  // Keep cf_handles_ after the client_ as column family handles need to be deleted before deleting the client. See
+  // CfDeleter.
+  std::shared_ptr<Client> client_;
+  std::map<std::string, CfUniquePtr> cf_handles_;
+  friend class WriteBatch;
+  friend class Iterator;
+};
+
+inline WriteBatch::WriteBatch(const std::shared_ptr<const NativeClient> &client) : client_{client} {}
+
+template <typename KeySpan, typename ValueSpan>
+void WriteBatch::put(const std::string &cFamily, const KeySpan &key, const ValueSpan &value) {
+  auto s = batch_.Put(client_->columnFamilyHandle(cFamily), NativeClient::toSlice(key), NativeClient::toSlice(value));
+  NativeClient::throwOnError("batch put failed"sv, std::move(s));
+}
+
+template <typename KeySpan, typename ValueSpan>
+void WriteBatch::put(const KeySpan &key, const ValueSpan &value) {
+  return put(NativeClient::defaultColumnFamily(), key, value);
+}
+
+template <typename KeySpan>
+void WriteBatch::del(const std::string &cFamily, const KeySpan &key) {
+  auto s = batch_.Delete(client_->columnFamilyHandle(cFamily), NativeClient::toSlice(key));
+  NativeClient::throwOnError("batch del failed"sv, std::move(s));
+}
```

Comment:
Do we expect that a client of this class will rarely give us to delete keys that do not exist, or it can actually be a frequent scenario?

and how should we behave when catching such exception? 
Do we need and have a way to differentiate it from other exceptions?  

I'm trying to understand if exceptions will complicate our implementation on scenarios that might be frequent

Reply:
Deleting a key that doesn't exist is not an error for both single-key deletes and batches. I've put a comment on the methods and added tests - `del_non_existent_key_in_batch_is_not_an_error` and `del_non_existent_key_is_not_an_error`. This is identical to RocksDB's behaviour.